source/util/move_to_front.cpp

```diff
@@ -0,0 +1,336 @@
+  const size_t kTextFieldWidthWithTimestamp = 14;
+  const size_t text_field_width = print_timestamp ?
+      kTextFieldWidthWithTimestamp : kTextFieldWidthWithoutTimestamp;
+
+  std::stringstream label;
+  label << IdOf(node) << "H" << HeightOf(node) << "S" << SizeOf(node);
+  if (print_timestamp)
+    label << "T" << TimestampOf(node);
+  const size_t label_length = label.str().length();
+  if (label_length < text_field_width)
+    label << std::string(text_field_width - label_length, '-');
+
+  out << label.str();
+
+  PrintTreeInternal(out, RightOf(node), depth + 1, print_timestamp);
+
+  if (LeftOf(node)) {
+    out << std::string(depth * text_field_width, ' ');
+    PrintTreeInternal(out, LeftOf(node), depth + 1, print_timestamp);
+  }
+}
+
+void MoveToFront::InsertNode(uint16_t node) {
+  if (!root_) {
+    root_ = node;
+    return;
+  }
+
+  uint16_t iter = root_;
+  uint16_t parent = 0;
+
+  bool right_child;
+
+  while (iter) {
+    parent = iter;
+    assert(TimestampOf(iter) != TimestampOf(node));
+    right_child = TimestampOf(iter) > TimestampOf(node);
+    iter = right_child ? RightOf(iter) : LeftOf(iter);
+  }
+
+  assert(parent);
+
+  MutableParentOf(node) = parent;
+
+  if (right_child)
+    MutableRightOf(parent) = node;
+  else
+    MutableLeftOf(parent) = node;
+
+  // Insertion is finished. Start the balancing process.
+  bool needs_rebalancing = true;
+  parent = ParentOf(node);
+
+  while (parent) {
+    UpdateNode(parent);
+
+    if (needs_rebalancing) {
+      const int parent_balance = BalanceOf(parent);
+
+      if (RightOf(parent) == node) {
+        // Added node to the right subtree.
+        if (parent_balance > 1) {
+          // Parent is right heavy, rotate left.
+          if (BalanceOf(node) < 0)
+            RotateRight(node);
+          parent = RotateLeft(parent);
+        } else if (parent_balance == 0 || parent_balance == -1) {
+          // Parent is balanced or left heavy, no need to balance further.
+          needs_rebalancing = false;
+        }
+      } else {
+        // Added node to the left subtree.
+        if (parent_balance < -1) {
+          // Parent is left heavy, rotate right.
+          if (BalanceOf(node) > 0)
+            RotateLeft(node);
+          parent = RotateRight(parent);
+        } else if (parent_balance == 0 || parent_balance == 1) {
+          // Parent is balanced or right heavy, no need to balance further.
+          needs_rebalancing = false;
+        }
+      }
+    }
+
+    assert(BalanceOf(parent) >= -1 && (BalanceOf(parent) <= 1));
+
+    node = parent;
+    parent = ParentOf(parent);
+  }
+}
+
+uint16_t MoveToFront::RemoveNode(uint16_t node) {
+  // Instead of removing the |node| find another node which is easier to
+  // remove and swap them.
+  if (const uint16_t scapegoat = RightestDescendantOf(LeftOf(node))) {
+    std::swap(MutableIdOf(node), MutableIdOf(scapegoat));
+    std::swap(MutableTimestampOf(node), MutableTimestampOf(scapegoat));
```

Comment:
When I originally read the code I missed the LeftOf part.  So makes sense that the scapegoat can take the place of the removed node.

But I think symmetry argues that the leftest-of-the-right-child would be equally acceptable.   Minor point.